src/tactic/norm_cast.lean

```diff
@@ -25,225 +25,314 @@ try_for 1000 (mk_instance e)
-used to normalize casts.
+mk_simp_attribute push_cast "The `push_cast` simp attribute uses `squash_cast` and `move_cast` lemmas in the \"forward\" direction,
+to move casts toward the leaf nodes of the expression."
+
+private meta def aux_after_set (tac : expr → tactic (expr × (expr → expr))) :
+  expr → tactic (expr × (expr → expr))
+| (pi n bi d b) := do
+  uniq_n ← mk_fresh_name,
+  let b' := b.instantiate_var (local_const uniq_n n bi d),
+  (b', f) ← aux_after_set b',
+  return $ (
+    pi n bi d $ b'.abstract_local uniq_n,
+    λ e, lam n bi d $ ( f $ e (local_const uniq_n n bi d) ).abstract_local uniq_n
+  )
+| ty := tac ty
 
-Let r be = or ↔, then compositional lemmas of the shape
-Π ..., ↑(P a1 ... an) r P ↑a1 ... ↑an should be given the
-attribute move_cast.
+/-- Called after the `move_cast` attribute is applied to a declaration. -/
+private meta def move_cast_after_set (decl : name) (prio : ℕ) (pers : bool) : tactic unit :=
+do
+  (declaration.thm n l ty e) ← get_decl decl | failed,
+  let tac := λ ty, (flip_eq ty <|> flip_iff ty),
+  (ty', f) ← aux_after_set tac ty,
+  let e' := task.map f e,
+  let n' := new_name n,
+  add_decl (declaration.thm n' l ty' e'),
+  simp_attr.push_cast.set decl () tt
+
+/--
+This is an attribute of the norm_cast tactic.
+It's intended for lemmas of the following shapes:
+  - Π ..., ↑(P a1 ... an) = P ↑a1 ... ↑an
+  - Π ..., ↑(P a1 ... an) ↔ P ↑a1 ... ↑an
 -/
 @[user_attribute]
 meta def move_cast_attr : user_attribute simp_lemmas :=
 {
-    name      := `move_cast,
-    descr     := "attribute for lemmas of the shape Π ..., ↑(P a1 ... an) = P ↑a1 ... ↑an",
-    after_set := some after_set,
-    cache_cfg :=
-        { mk_cache     := mk_cache ∘ (list.map new_name),
-          dependencies := [], },
+  name      := `move_cast,
+  descr     := "attribute for lemmas of the shape Π ..., ↑(P a1 ... an) = P ↑a1 ... ↑an",
+  after_set := some move_cast_after_set,
+  cache_cfg :=
+    { mk_cache     := mk_cache ∘ (list.map new_name),
+      dependencies := [], },
 }
 
+/-
+This is an auxiliary function to merge the sets of elim_cast and move_cast lemmas,
+as they are used together in step 2.
+-/
 private meta def get_cache : tactic simp_lemmas :=
 do
-    a ← elim_cast_attr.get_cache,
-    b ← move_cast_attr.get_cache,
-    return $ simp_lemmas.join a b
+  a ← elim_cast_attr.get_cache,
+  b ← move_cast_attr.get_cache,
+  return $ simp_lemmas.join a b
 
-/--
-This is an attribute for simplification rules of the shape
-Π ..., ↑↑a = ↑a or  Π ..., ↑a = a.
+/-- Called after the `squash_cast` attribute is applied to a declaration. -/
+private meta def squash_cast_after_set (decl : name) (prio : ℕ) (pers : bool) : tactic unit :=
+simp_attr.push_cast.set decl () tt
 
-They are used in a heuristic to infer intermediate casts.
+/--
+This is an attribute of the norm_cast tactic for simplification rules that compose or
+remove casts, but without generalizing the expression in the way elim_cast lemmas do.
```

Comment:
I don't really understand "generalizing the expression" here. This is related to my questions above.

Reply:
This should mirror the description in the documentation, I forgot to update it at some point. I've had a lot of trouble getting my mind around how squash_cast should be defined.
This definitely doesn't make a lot of sense, but I think what I meant is that rewriting using a squash_cast lemma doesn't make any progress, it doesn't move a cast past an operator for instance. If you put a squash_cast lemma in move_cast or elim_cast, norm_cast will diverge. For instance it will rewrite 1 to coe 1 to coe (coe 1), ...